// object-gui/dist/js/Editors/EditorStringInput.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./EditorInputBase"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.EditorStringInput = void 0;
    const EditorInputBase_1 = require("./EditorInputBase");
    class EditorStringInput extends EditorInputBase_1.EditorInputBase {
        constructor(data, id, name, value, autoMap = true) {
            super(data, () => document.createElement("input"), id, name, () => "", (v) => v, (self) => {
                const inputEl = self.element;
                return inputEl.value;
            }, (self, v) => {
                const inputEl = self.element;
                inputEl.value = v;
            }, value, autoMap);
            const input = this.element;
            input.type = "text";
            input.addEventListener("change", () => {
                this.changeEventHandler();
            });
        }
        addItem() {
            return this;
        }
        addItemGroup() {
            return this;
        }
        addItems() {
            return this;
        }
        max() {
            return this;
        }
        min() {
            return this;
        }
        step() {
            return this;
        }
    }
    exports.EditorStringInput = EditorStringInput;
});
