// object-gui/dist/js/Editors/index.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./EditorButton", "./EditorCheckboxInput", "./EditorColorInput", "./EditorGroup", "./EditorItem", "./EditorNumberInput", "./EditorSelectInput", "./EditorStringInput"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.EditorStringInput = exports.EditorSelectInput = exports.EditorNumberInput = exports.EditorItem = exports.EditorGroup = exports.EditorColorInput = exports.EditorCheckboxInput = exports.EditorButton = void 0;
    var EditorButton_1 = require("./EditorButton");
    Object.defineProperty(exports, "EditorButton", { enumerable: true, get: function () { return EditorButton_1.EditorButton; } });
    var EditorCheckboxInput_1 = require("./EditorCheckboxInput");
    Object.defineProperty(exports, "EditorCheckboxInput", { enumerable: true, get: function () { return EditorCheckboxInput_1.EditorCheckboxInput; } });
    var EditorColorInput_1 = require("./EditorColorInput");
    Object.defineProperty(exports, "EditorColorInput", { enumerable: true, get: function () { return EditorColorInput_1.EditorColorInput; } });
    var EditorGroup_1 = require("./EditorGroup");
    Object.defineProperty(exports, "EditorGroup", { enumerable: true, get: function () { return EditorGroup_1.EditorGroup; } });
    var EditorItem_1 = require("./EditorItem");
    Object.defineProperty(exports, "EditorItem", { enumerable: true, get: function () { return EditorItem_1.EditorItem; } });
    var EditorNumberInput_1 = require("./EditorNumberInput");
    Object.defineProperty(exports, "EditorNumberInput", { enumerable: true, get: function () { return EditorNumberInput_1.EditorNumberInput; } });
    var EditorSelectInput_1 = require("./EditorSelectInput");
    Object.defineProperty(exports, "EditorSelectInput", { enumerable: true, get: function () { return EditorSelectInput_1.EditorSelectInput; } });
    var EditorStringInput_1 = require("./EditorStringInput");
    Object.defineProperty(exports, "EditorStringInput", { enumerable: true, get: function () { return EditorStringInput_1.EditorStringInput; } });
});
